### PR TITLE
Example base-gtkmm

### PR DESCRIPTION
**Pull request by JacksonCampolattaro**
Created a new example that demonstrates how to integrate magnum's graphics handling into a Gtkmm interface. The example contains a class `MagnumWidget` which extends Gtkmm's own `GLArea` widget. In the example, the graphics widget is the only thing shown in the window, but in actual use it could be embedded as a viewport as part of a larger application. Comments provide instructions for where to add preparation/drawing/cleanup code.

### CODE REVIEW

File: src/MyApplication.cpp [hunk 1]
```diff
@@ -1,67 +1,109 @@
...
+#include <gtkmm/window.h>
+#include <gtkmm/glarea.h>
 
 using namespace Magnum;
 
-class MyApplication: public QOpenGLWidget {
-    public:
-        explicit MyApplication(Platform::GLContext& context, QWidget* parent = nullptr, Qt::WindowFlags f = nullptr);
+class MagnumWidget : public Gtk::GLArea {
+public:
```
Comment (mosra) [inline src/MyApplication.cpp]: Can you keep the indentation of the `public`/`private` labels as it was before?

File: src/MyApplication.cpp [hunk 2]
```diff
@@ -1,67 +1,109 @@
...
+    void onResize(int width, int height);
+    void onUnrealize();
 
-        Platform::GLContext& _context;
+    Platform::GLContext &_context;
 };
 
-MyApplication::MyApplication(Platform::GLContext& context, QWidget* parent, Qt::WindowFlags f): QOpenGLWidget{parent, f}, _context(context) {
-    /* TODO: Add your context format setup code here */
+MagnumWidget::MagnumWidget(Platform::GLContext &context) : Gtk::GLArea(), _context(context) {
```
Comment (mosra) [inline src/MyApplication.cpp]:
```suggestion
MagnumWidget::MagnumWidget(Platform::GLContext& context): Gtk::GLArea{}, _context{context} {
```

In particular, `&` (and `*`) goes next to the type, no space in front of `:`, prefer using `{}` over `()` for constructors. Same elsewhere.

Btw., is the explicit call to `Gtk::GLArea` default constructor needed? I think not.

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -13,14 +13,17 @@ else()
     message(FATAL_ERROR "Magnum context creation is not supported on this platform")
 endif()
 
-set_directory_properties(PROPERTIES CORRADE_USE_PEDANTIC_FLAGS ON)
+
+#set_directory_properties(PROPERTIES CORRADE_USE_PEDANTIC_FLAGS ON)
```
Comment (mosra) [inline src/CMakeLists.txt]: Why is this commented out? Was it causing a lot of warnings printed on the output?

In that case, maybe `include_directories(SYSTEM ${GTKMM_INCLUDE_DIRS})` could help.

I tried to find a "modern CMake" way of finding GTK, but sadly there's nothing ready-made and the official guide recommends using pkg-config, which isn't really portable. Oh well.

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -13,14 +13,17 @@ else()
...
 add_executable(MyApplication MyApplication.cpp)
 target_link_libraries(MyApplication PRIVATE
-    Magnum::GLContext
-    Magnum::GL
-    Magnum::Magnum
-    Qt5::Widgets)
+        Magnum::GLContext
+        Magnum::GL
+        Magnum::Magnum
+        ${GTKMM_LIBRARIES})
```
Comment (mosra) [inline src/CMakeLists.txt]: Indent consecutive lines with just four spaces, please :)

File: CMakeLists.txt [hunk 5]
```diff
@@ -3,4 +3,7 @@ project(MyApplication)
 
 set(CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/modules/" ${CMAKE_MODULE_PATH})
 
+find_package(PkgConfig REQUIRED)
+pkg_check_modules(GTKMM gtkmm-3.0)
```
Comment (mosra) [inline CMakeLists.txt]: This could be inside `src/CMakeLists.txt` to keep it localized.

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -1,6 +1,6 @@
-find_package(Qt5Widgets REQUIRED)
 find_package(Magnum REQUIRED GL)
 
+
```
Comment (mosra) [inline src/CMakeLists.txt]: Style: no more than one consecutive empty line, please :) Same elsewhere.

File: src/MyApplication.cpp [hunk 7]
```diff
@@ -1,67 +1,109 @@
...
+    /* Retrieve the ID of the relevant framebuffer */
+    GLint framebufferID;
+    glGetIntegerv(GL_FRAMEBUFFER_BINDING, &framebufferID);
 
-    qtDefaultFramebuffer.clear(GL::FramebufferClear::Color);
+    /* Attache Magnum's framebuffer manager to the framebuffer provided by Gtkmm */
+    auto gtkmmDefaultFramebuffer = GL::Framebuffer::wrap(framebufferID, {{}, {get_width(), get_height()}});
+
+    /* Reset state to avoid Gtkmm affecting Magnum */
+    GL::Context::current().resetState(GL::Context::State::ExitExternal);
```
Comment (mosra) [inline src/MyApplication.cpp]: I think this line should be the first thing in the function, to be safe. It won't affect the `glGet` call in any way.

File: src/CMakeLists.txt (hunk 3 above)
Comment (JacksonCampolattaro) [reply src/CMakeLists.txt]: Yeah, that's exactly what was happening, thanks for the tip! It's a shame pkg-config is necessary, I'm hoping they update that when Gtkmm 4 comes out.

File: CMakeLists.txt (hunk 5 above)
Comment (JacksonCampolattaro) [reply CMakeLists.txt]: I thought it was odd that the official guide recommends putting it in the top level cmake file, too. I'm pretty sure it works either way, so I'll move it.

**Comment by mosra** [comment]
Merged as 12f4da7a143db1c9460d9f6c04fe9c9c222484b1 (and the usage instructions as b7423e706b53e2547829b8516ac5de52e457f009). Thank you!

**Pull request closed without merging**